Important bug fixes, docs, IR spectrographs

jhennawi opened the pull request:
This PR fixes bugs that were introduced by the last load_save_specobj PR that were only discovered after careful analysis of the outputs. Not only was coadd2d broken, but echelle object finding had a subtle but catastrophic object indexing bug.  Since develop is basically currently compromised, this should be approved right away. Tests are running right now. 

There are other things in progress in this PR related to IR spectrographs that are being introduced, fixes to 2d coadding, and documentation.  Given parallel development efforts (i.e. load_save_specobjs and my IR spectrograph PR), there was no simple way to fix these bugs without creating lots of conflicts, so this stuff comes along for the ride for now.

jhennawi commented:
Dev suite passes all tests. 

<img width="808" alt="Screen Shot 2019-11-28 at 12 20 15" src="https://user-images.githubusercontent.com/8774563/69830392-8da1fb80-11d9-11ea-9ffc-0df3bb7c7ac4.png">

jhennawi commented:
Tests passed. Cooked copied over. This one is ready to be approved. 

<img width="906" alt="Screen Shot 2019-11-28 at 13 39 59" src="https://user-images.githubusercontent.com/8774563/69832727-b54a9100-11e4-11ea-8499-1c45d32e5f12.png">

profxj commented:
@jhennawi -- Please put in all the docs needed for `develop` to go into `master` in this PR as originally planned.  I'll then review.  Thanks.

jhennawi commented:
@profxj this PR is urgent because your PR does not pass coadd2d, and it also contains a bunch of bugs which I've fixed in this PR. For example, your copy method was using pointers and not making copies, wreaking havoc. Develop is not currently fully functional, so I recommend you review and approve this PR first. I'm working on the docs for those codes, but the they also now need to be reworked and fixed, because your PR broke most of the  new coadding, fluxing, and telluric routines. People using develop for those purposes (i.e. anyone doing IR reductions) are now dead in the water, because they cannot combine, flux, or telluric correct their data. I should have admittedly guarded those routines with tests so this is certainly partly my fault, but please also coordinate better when making major changes to the data model as you did with specobjs.

kbwestfall commented on pypeit/core/pydl.py:
```diff
@@ -652,7 +653,11 @@ def value(self, x, x2=None, action=None, lower=None, upper=None):
             goodcoeff = self.coeff[:, coeffbk]
         else:
-            goodcoeff = self.coeff[coeffbk]
+            try:
```
remove try block?

jhennawi replied:
Same point as above. Will clean it up before we merge.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -4155,8 +4158,11 @@ def convert_to_tslits_dict(self):
+        # JFH This is causing problems and I cannot figure out how to toggle is_sync when creating psuedo masters. Let's
+        # assume if we are converting to a tslits_dict we are always already synced.
+        #if not self.is_synced:
+        #    msgs.error('Edges must be synced to construct tslits_dict.')
```
`is_synced` is a direct check of the object itself; it's not a boolean that can be toggled.  If this is being tripped, it means that the object isn't synced and the output of this method is likely to be wrong in some way.  There are some annoying corner cases where the `sync` method fails and the result is that it's not synced.  If you point me to data where this was causing you problems, I can help you figure out why.  But this check needs to be in place for the tslits_dict conversion to be trustworthy.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -4220,6 +4228,8 @@ def from_tslits_dict(cls, tslits_dict, master_key, master_dir):
         this.spat_err = np.zeros(this.spat_cen.shape, dtype=float)
         this.spat_img = np.round(this.spat_cen).astype(int)
+        # JFH This does not work.
+        #this.is_synced = True
```
It's not a toggle.  If the instance created by the `from_tslits_dict` isn't synced, there may be a problem in how it's being setup. If so, I can help figure out why.

kbwestfall commented on pypeit/io.py:
```diff
@@ -187,7 +188,13 @@ def parse_hdr_key_group(hdr, prefix='F'):
+    try:
+        return [values[i] for i in range(max(values.keys())+1)]
+    except:
+        return []
```
Would probably be good to add this as an issue.  Or is it one already that you can reference in the comment?

jhennawi replied:
Issued added.

kbwestfall commented on pypeit/scripts/show_2dspec.py:
```diff
@@ -40,6 +38,8 @@ def parser(options=None):
     parser.add_argument('--showmask', default=False, help='Overplot masked pixels',
                         action='store_true')
+    parser.add_argument('--clean', default=False, help="Remove trace plotting",
+                        action = "store_true")
```
Just an opinion, but `clean` feels non-intuitive.  As a user, I'm not sure I would know what you mean by "Remove" trace plotting.  It's just that you're not overplotting the trace data, right?

feigewang replied:
changed clean to removetrace to make it clearer

kbwestfall commented on pypeit/scripts/show_2dspec.py:
```diff
@@ -171,32 +172,32 @@ def main(args):
-    if sobjs is not None:
-        show_trace(sobjs, args.det, viewer, ch)
+    if not args.clean:
+        if sobjs is not None:
```
` if not args.clean and sobjs is not None:`

feigewang replied:
Thanks. fixed

kbwestfall commented on pypeit/scripts/show_2dspec.py:
```diff
@@ -171,32 +172,32 @@ def main(args):
-    if sobjs is not None:
-        show_trace(sobjs, args.det, viewer, ch)
+    if not args.clean:
+        if sobjs is not None:
```
`if not args.clean and sobjs is not None:`

feigewang replied:
fixed

kbwestfall commented on pypeit/scripts/show_2dspec.py:
```diff
@@ -171,32 +172,32 @@ def main(args):
-        show_trace(sobjs, args.det, viewer, ch)
+                                  cuts = (-5.0, 5.0), bitmask = mask_in)
+    if not args.clean:
+        if sobjs is not None:
```
`if not args.clean and sobjs is not None:`

feigewang replied:
fixed

kbwestfall commented on pypeit/specobj.py:
```diff
@@ -80,16 +80,23 @@
     'VEL_CORR': dict(otype=float, desc='Relativistic velocity correction for wavelengths'),
     #
-    'DET': dict(otype=(int,np.int64), desc='Detector number'),
+    'DET': dict(otype=(int,np.int64,np.int32), desc='Detector number'),
```
Use `(int, np.integer)` instead?

E.g.:
```
>>> x = numpy.int64(3)
>>> y = numpy.int32(3)
>>> z = 3
>>> type(z)
<class 'int'>
>>> isinstance(x, numpy.integer)
True
>>> isinstance(y, numpy.integer)
True
>>> isinstance(z, numpy.integer)
False
```

jhennawi replied:
Is there an analog for this for float 32 and float 64. np.float does not work??

kbwestfall replied:
Yeah, `np.floating`.

```
>>> import numpy
>>> x = numpy.float64(2.2)
>>> y = numpy.float32(2.2)
>>> z = 2.2
>>> type(z)
<class 'float'>
>>> isinstance(x, numpy.floating)
True
>>> isinstance(y, numpy.floating)
True
>>> isinstance(z, numpy.floating)
False
```

kbwestfall commented on pypeit/spectrographs/keck_mosfire.py:
```diff
@@ -0,0 +1,244 @@
+        return par
+
+
+#    def default_pypeit_par():
```
remove?

jhennawi replied:
Done.

kbwestfall commented on pypeit/spectrographs/keck_nirspec.py:
```diff
@@ -45,44 +45,113 @@ def __init__(self):
+    def default_pypeit_par(self):
         """
-        Set default parameters for NIRSPEC reductions
+        Set default parameters for Keck/MOSFIRE
```
Think this is a typo.

jhennawi replied:
fixed.

kbwestfall commented on pypeit/spectrographs/keck_nirspec.py:
```diff
@@ -45,44 +45,113 @@ def __init__(self):
+        par['scienceframe']['exprng'] = [20, None]
         return par
 
+    # JFH Replaced with values from MOSFIRE above.
```
Not sure what happened here, but can you check that these changes were intentional?

jhennawi replied:
Yeah it was, but I'm keeping the old setting sin there for now.

kbwestfall commented on pypeit/spectrographs/keck_nirspec.py:
```diff
@@ -244,7 +313,7 @@ def lamps(self, fitstbl, status):
 
     # TODO: This function is unstable to shape...
-    def bpm(self, filename, det):
+    def bpm(self, filename, det, shape=None):
```
is this correct?

jhennawi replied:
I think it is now, see also analogous algorithm in spectrograph.

jhennawi commented on pypeit/core/arc.py:
```diff
@@ -1040,11 +1040,11 @@ def fit_arcspec(xarray, yarray, pixt, fitp):
-    cent   = -1.0*np.ones(sz_p, dtype=np.float)
-    widt   = -1.0*np.ones(sz_p, dtype=np.float)
-    centerr = -1.0*np.ones(sz_p, dtype=np.float)
+    b      = -999.0*np.ones(sz_p, dtype=np.float)
```
Hi @kbwestfall. Thanks.  Run the LRIS 2-d and GNIRS 2-d coadd on this branch and then try to read in the objects with pypeit_show_2dspec. That is where I discovered the problem with is_synced. So put that line of code back in, and that is where the problem arose.

jhennawi replied:
changed.

jhennawi commented on pypeit/core/coadd2d.py:
```diff
@@ -568,6 +580,79 @@ def __init__(self, spec2d_files, spectrograph, det=1, offsets=None, weights='aut
+        Args:
+            slitorderid (int):
+               The slit or order id that has the brightest object whose S/N will be used to determine the weight for each frame.
+            objid (np.ndarray): shape = (nexp,)
```
Okay will move the shape to the description.

jhennawi commented on pypeit/core/coadd2d.py:
```diff
@@ -568,6 +580,79 @@ def __init__(self, spec2d_files, spectrograph, det=1, offsets=None, weights='aut
+               Use constant weights for coadding the exposures. Default=False
+
+        Returns:
+            rms_sn, weights
```
Okay I'll give it a go.

jhennawi replied:
Can you have a look at this now and tell me if it is correct. I just deleted the first line.

kbwestfall replied:
No, this still won't get formatted correctly.  Here's what I would write:

```
    Returns:
        tuple: Two numpy.ndarray objects are returned, both have shape
        ``(len(specobjs_list),)``:

            - rms_sn: The root mean square S/N value for each input
              spectrum.
            - weights: The signal-to-noise-squared weights to be applied
              to the spectra.

```

kbwestfall commented:
I fixed the problem with the `is_synced` check.  It was a bug in how the `EdgeTraceSet` instance was being instantiated from the `tslits_dict`.  I've confirmed that the call to `pypeit_show_2dspec` now works with GNIRS.

On another note, your run of the dev suite shows 31 tests, but aren't there 33 now?

jhennawi commented on pypeit/core/extract.py:
```diff
@@ -513,7 +514,10 @@ def fit_profile(image, ivar, waveimg, thismask, spat_img, trace_in, wave, flux,
                                      kwargs_bspline={'everyn': 30}, kwargs_reject={'groupbadpix':True,'maxrej':1})
     spline_flux, _ = b_answer.value(wave[indsp])
-    cont_flux, _ = c_answer.value(wave[indsp])
+    try:
```
This is an extremely rare bug that I'm still fixing. Leaving it now so I don't forget.

jhennawi commented on pypeit/core/flat.py:
```diff
@@ -472,6 +472,9 @@ def fit_flat(flat, tilts_dict, tslits_dict_in, slit, inmask = None,
     pixelflat[flat >= nonlinear_counts] = 1.0
+    # Do not apply pixelflat field corrections that are greater than 100% to avoid creating edge effects, etc.
+    # TODO Should we do the same for the illumflat??
+    pixelflat = np.fmax(np.fmin(pixelflat, 2.0), 0.5)
```
Changed.

jhennawi commented:
Merging
